File: frontend/studi-app/utils/dateUtils.ts
// Date utilities for dashboard navigation

export function formatDateForDisplay(date: Date, type: 'daily' | 'weekly'): string {
  const today = new Date();
  const yesterday = new Date(today);
  yesterday.setDate(today.getDate() - 1);
  
  if (type === 'daily') {
    // Check if it's today
    if (isSameDay(date, today)) {
      return 'Today';
    }
    // Check if it's yesterday
    if (isSameDay(date, yesterday)) {
      return 'Yesterday';
    }
    
    // Format as "Thursday, Jan 23" or "Thursday, Jan 23, 2024" if different year
    const options: Intl.DateTimeFormatOptions = {
      weekday: 'long',
      month: 'short',
      day: 'numeric',
    };
    
    if (date.getFullYear() !== today.getFullYear()) {
      options.year = 'numeric';
    }
    
    return date.toLocaleDateString('en-US', options);
  } else {
    // Weekly format
    const currentWeekStart = getWeekStart(today);
    const lastWeekStart = new Date(currentWeekStart);
    lastWeekStart.setDate(currentWeekStart.getDate() - 7);
    
    const weekStart = getWeekStart(date);
    
    // Check if it's this week
    if (isSameDay(weekStart, currentWeekStart)) {
      return 'This Week';
    }
    // Check if it's last week
    if (isSameDay(weekStart, lastWeekStart)) {
      return 'Last Week';
    }
    
    // Format as "Jan 13-19" or "Jan 13-19, 2024" if different year
    const weekEnd = new Date(weekStart);
    weekEnd.setDate(weekStart.getDate() + 6);
    
    const startMonth = weekStart.toLocaleDateString('en-US', { month: 'short' });
    const startDay = weekStart.getDate();
    const endDay = weekEnd.getDate();
    
    let formattedRange = `${startMonth} ${startDay}-${endDay}`;
    
    if (weekStart.getFullYear() !== today.getFullYear()) {
      formattedRange += `, ${weekStart.getFullYear()}`;
    }
    
    return formattedRange;
  }
}

export function isSameDay(date1: Date, date2: Date): boolean {
  return (
    date1.getFullYear() === date2.getFullYear() &&
    date1.getMonth() === date2.getMonth() &&
    date1.getDate() === date2.getDate()
  );
}

export function getWeekStart(date: Date): Date {
  const d = new Date(date);
  const day = d.getDay(); // 0 = Sunday, 1 = Monday, ..., 6 = Saturday

  const weekStart = new Date(d);
  // Convert to Monday-based week: Monday = 0, Tuesday = 1, ..., Sunday = 6
  const mondayOffset = day === 0 ? 6 : day - 1; // Sunday becomes 6, Monday becomes 0
  // Move backwards to get Monday
  weekStart.setDate(d.getDate() - mondayOffset);

  // Zero-out time portion for consistency
  weekStart.setHours(0, 0, 0, 0);
  return weekStart;
}

export function getWeekEnd(date: Date): Date {
  const weekStart = getWeekStart(date);
  const weekEnd = new Date(weekStart);
  weekEnd.setDate(weekStart.getDate() + 6);
  return weekEnd;
}

export function navigateDate(currentDate: Date, direction: 'prev' | 'next', type: 'daily' | 'weekly'): Date {
  const newDate = new Date(currentDate);
  
  if (type === 'daily') {
    newDate.setDate(currentDate.getDate() + (direction === 'next' ? 1 : -1));
  } else {
    newDate.setDate(currentDate.getDate() + (direction === 'next' ? 7 : -7));
  }
  
  return newDate;
}

export function canNavigate(currentDate: Date, direction: 'prev' | 'next', type: 'daily' | 'weekly' | 'monthly', isPremium: boolean = false): boolean {
  const today = new Date();
  
  if (direction === 'next') {
    // Can't navigate beyond today/current week/month (prevent future navigation)
    if (type === 'daily') {
      return currentDate < today;
    } else if (type === 'weekly') {
      const currentWeekStart = getWeekStart(today);
      const targetWeekStart = getWeekStart(currentDate);
      return targetWeekStart < currentWeekStart;
    } else { // monthly
      const currentMonthStart = getMonthStart(today);
      const targetMonthStart = getMonthStart(currentDate);
      return targetMonthStart < currentMonthStart;
    }
  } else {
    // Allow unlimited historical navigation - data access will be gated separately
    return true;
  }
}

export function formatDateForAPI(date: Date): string {
  // Keep the same calendar date but send it as a date string to server
  // This avoids timezone conversion issues while preserving the user's intended date
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
}

export function getDefaultDate(type: 'daily' | 'weekly'): Date {
  const today = new Date();
  if (type === 'weekly') {
    return getWeekStart(today);
  }
  return today;
}

// Returns an array of 7 Date objects representing the Mon→Sun week window
export function getWeekDays(weekStart: Date): Date[] {
  const days: Date[] = [];
  for (let i = 0; i < 7; i++) {
    const d = new Date(weekStart);
    d.setDate(weekStart.getDate() + i);
    days.push(d);
  }
  return days;
}

// Convenience for moving an entire week window backward / forward
export function navigateWeek(currentWeekStart: Date, direction: 'prev' | 'next'): Date {
  const newWeek = new Date(currentWeekStart);
  newWeek.setDate(currentWeekStart.getDate() + (direction === 'next' ? 7 : -7));
  return newWeek;
}

// Month utility functions
export function getMonthStart(date: Date): Date {
  const monthStart = new Date(date);
  monthStart.setDate(1);
  monthStart.setHours(0, 0, 0, 0);
  return monthStart;
}

export function getMonthEnd(date: Date): Date {
  const monthEnd = new Date(date);
  monthEnd.setMonth(monthEnd.getMonth() + 1);
  monthEnd.setDate(0);
  monthEnd.setHours(23, 59, 59, 999);
  return monthEnd;
} 

File: backend/analytics/services/date_utils.py
from datetime import date, timedelta
from django.utils import timezone


def get_week_boundaries(target_date):
    """
    Returns the Monday and Sunday dates for the week containing the given date (Mon→Sun)
    Args:
        target_date: date object
    Returns:
        tuple: (monday_date, sunday_date)
    """
    if isinstance(target_date, str):
        target_date = date.fromisoformat(target_date)
    
    # In Python's weekday(): Monday=0, Tuesday=1, ..., Sunday=6
    # We want the week to start on Monday and end on Sunday
    
    # Get the Monday of this week
    monday_offset = target_date.weekday()  # 0 for Monday, 1 for Tuesday, ..., 6 for Sunday
    monday = target_date - timedelta(days=monday_offset)
    
    # Sunday is 6 days after Monday
    sunday = monday + timedelta(days=6)

    return monday, sunday


def get_month_boundaries(target_date):
    """
    Returns the first and last dates of the month containing the given date
    Args:
        target_date: date object
    Returns:
        tuple: (month_start, month_end)
    """
    if isinstance(target_date, str):
        target_date = date.fromisoformat(target_date)
    
    # First day of the month
    month_start = target_date.replace(day=1)
    
    # Last day of the month
    if target_date.month == 12:
        next_month = target_date.replace(year=target_date.year + 1, month=1, day=1)
    else:
        next_month = target_date.replace(month=target_date.month + 1, day=1)
    month_end = next_month - timedelta(days=1)
    
    return month_start, month_end


def is_current_period(target_date, timeframe):
    """
    Check if the given date represents the current period for the timeframe
    Args:
        target_date: date object
        timeframe: 'daily', 'weekly', or 'monthly'
    Returns:
        bool: True if this is the current period
    """
    today = timezone.now().date()
    
    if timeframe == 'daily':
        return target_date == today
    elif timeframe == 'weekly':
        current_week_start, _ = get_week_boundaries(today)
        target_week_start, _ = get_week_boundaries(target_date)
        return current_week_start == target_week_start
    elif timeframe == 'monthly':
        current_month_start, _ = get_month_boundaries(today)
        target_month_start, _ = get_month_boundaries(target_date)
        return current_month_start == target_month_start
    
    return False 